office-365-feed/operations.py

```diff
@@ -0,0 +1,275 @@
+"""
+  Copyright start
+  Copyright (C) 2008 - 2022 Fortinet Inc.
+  All rights reserved.
+  FORTINET CONFIDENTIAL & FORTINET PROPRIETARY SOURCE CODE
+  Copyright end
+"""
+import requests
+import uuid
+import datetime
+import time
+import re
+from .constants import *
+from connectors.core.connector import get_logger, ConnectorError
+
+logger = get_logger('office-365-feed')
+
+
+def ip_to_indicator_type(ip):
+    # Returns the indicator type of the input IP.
+    if re.match(ipv4cidrRegex, ip):
+        return "IPv4 Address"
+
+    elif re.match(ipv4Regex, ip):
+        return "IPv4 Address"
+
+    elif re.match(ipv6cidrRegex, ip):
+        return "IPv6 Address"
+
+    elif re.match(ipv6Regex, ip):
+        return "IPv6 Address"
+
+    else:
+        return None
+
+
+def check_indicator_type(indicator):
+    # Checks the indicator type.
+    is_ip_indicator = ip_to_indicator_type(indicator)
+    if is_ip_indicator:
+        return is_ip_indicator
+    elif '*' in indicator:
+        return "URL"
+    else:
+        return "URL"
+
+
+def convert_str_list(params):
+    if not params:
+        return []
+    param_list = list(map(lambda x: x.strip(' '), params.split(','))) if isinstance(params, str) else params
+    return param_list
+
+
+def get_datetime(_epoch):
+    pattern = '%Y-%m-%dT%H:%M:%S.%fZ'
+    return str(datetime.datetime.utcfromtimestamp(_epoch).strftime(pattern))
+
+
+def max_age(params):
+    if params.get("expiry") is not None and params.get("expiry") != '':
+        return int(time.time()) + (params.get("expiry") * 86400)
+    else:
+        # Setting the default expiry time to 15 days from the indicator fetch time.
+        return int(time.time()) + (15 * 86400)
+
+
+def build_urls_dict(regions_list, services_list, unique_id, endpoint, latest_version_list=[]):
+    try:
+        urls_list = []
+        service = ','.join(map(str, services_list))
+        for region in regions_list:
+            if service == 'All':
+                url = 'https://endpoints.office.com/{0}/{1}?ClientRequestId={2}'.format(endpoint, region, unique_id)
+            else:
+                url = 'https://endpoints.office.com/{0}/{1}?ServiceAreas={2}' \
+                      '&ClientRequestId={3}'.format(endpoint, region, service, unique_id)
+
+            if latest_version_list:
+                for dict_ in [x for x in latest_version_list if x["instance"] == region]:
+                    urls_list.append({'lastUpdated': dict_.get("latest"),
```

Comment:
will we need lastUpdated in config also now that it is being stored in macro?

Reply:
ln this case lastUpdated is pulled from API to check the latest updated time for indicators.  We always fetch the latest time and compare it with PB macro. So not required to store it in config.